runtime/extern.c

```diff
@@ -536,15 +536,35 @@ static void extern_rec(value v)
       continue;
     case Custom_tag: {
       uintnat sz_32, sz_64;
+      char * size_header;
       char * ident = Custom_ops_val(v)->identifier;
       void (*serialize)(value v, uintnat * bsize_32,
                         uintnat * bsize_64)
         = Custom_ops_val(v)->serialize;
+      const struct custom_fixed_length* fixed_length
+        = Custom_ops_val(v)->fixed_length;
       if (serialize == NULL)
         extern_invalid_argument("output_value: abstract value (Custom)");
-      write(CODE_CUSTOM);
-      writeblock(ident, strlen(ident) + 1);
-      Custom_ops_val(v)->serialize(v, &sz_32, &sz_64);
+      if (!fixed_length) {
+        write(CODE_CUSTOM_LEN);
+        writeblock(ident, strlen(ident) + 1);
+        /* Reserve 12 bytes for the lengths (sz_32 and sz_64). */
+        if (extern_ptr + 12 >= extern_limit) grow_extern_output(12);
+        size_header = extern_ptr;
+        extern_ptr += 12;
+        Custom_ops_val(v)->serialize(v, &sz_32, &sz_64);
+        /* Store length before serialized block */
+        store32(size_header, sz_32);
+        store64(size_header + 4, sz_64);
+      } else {
+        write(CODE_CUSTOM_FIXED);
+        writeblock(ident, strlen(ident) + 1);
+        Custom_ops_val(v)->serialize(v, &sz_32, &sz_64);
+        if (sz_32 != fixed_length->bsize_32 ||
+            sz_64 != fixed_length->bsize_64)
+          caml_fatal_error("output_value: incorrect fixed sizes specified by %s",
```

Comment:
I'm tempted to put the check in a `CAMLassert`, so that it is performed by the debug runtime but not by the normal runtime.

Reply:
It uses `caml_fatal_error` so that it can specify the name of the offending serialiser (an assertion failure is harder to debug, and looks like a bug in the runtime rather than the user's custom serialiser)